### ISSUE TITLE
User not notified about session invitation failure when an error occurs.

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#674]
Users are not notified that the session negotiation failed because of a version mismatch (the Saros version of the negotiation participants were not compatible). Instead, the negotiation just fails silently.

A better behavior would be to show a notification to the user stating this fact.

**Comment by tobous** [comment]
I wasn't able to find the matching logic that deals with this on the Eclipse side. It doesn't seem to be a cancellation handler (as no such handler is registered with the outgoing session negotiation). It also can't be solely based on the status as the message shown to the user contains the exception thrown in [OutgoingSessionNegotiation.java#L258](https://github.com/saros-project/saros/blob/master/core/src/saros/negotiation/OutgoingSessionNegotiation.java#L258).

@srossbach Could you provide some more insight on how this is handled in Saros/E?

**Comment by srossbach** [comment]
https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/context/SarosIntellijContextFactory.java#L98

This is the version number that was used in the first release. I guess it is still the same. The method says that it returns the SDK version which is just garbage.

**Comment by srossbach** [comment]
Eclipse version is 

https://github.com/saros-project/saros/blob/master/eclipse/src/saros/SarosEclipseContextFactory.java#L129

This returns the version you entered in the plugin.xml file.

**Comment by tobous** [comment]
> https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/context/SarosIntellijContextFactory.java#L98
> 
> This is the version number that was used in the first release. I guess it is still the same. The method says that it returns the SDK version which is just garbage.

I am not completely sure what you mean, but I am pretty sure that the correct version numbers are read from the plugin configuration file and used (see stacktrace below). The logic is defined in the [IntelliJVersionProvider.java](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/context/IntelliJVersionProvider.java#L23).
```Java
ERROR 21:07:48,654 [Inviting intellij_test_2@saros-con.imp.fu-berlin.de...] (OutgoingSessionNegotiation.java:257) OSN [remote side: intellij_test_2@saros-con.imp.fu-berlin.de/Saros] : Saros versions are not compatible
DEBUG 21:07:48,657 [Inviting intellij_test_2@saros-con.imp.fu-berlin.de...] (Negotiation.java:221) negotiation OSN [remote side: intellij_test_2@saros-con.imp.fu-berlin.de/Saros] was canceled by the local side, error: The Saros plugin of intellij_test_2@saros-con.imp.fu-berlin.de (Version 0.1.0) is not compatible with your installed Saros plugin (Version 0.2.0)
ERROR 21:07:48,658 [Inviting intellij_test_2@saros-con.imp.fu-berlin.de...] (Negotiation.java:438) canceled negotiation OSN [remote side: intellij_test_2@saros-con.imp.fu-berlin.de/Saros], error: The Saros plugin of intellij_test_2@saros-con.imp.fu-berlin.de (Version 0.1.0) is not compatible with your installed Saros plugin (Version 0.2.0)
```

My question was not regarding how the version numbers where read/provided but rather how the cancellation dialog was shown on the host side when a mismatch is detected and the session negotiation is aborted.

This dialog is present in Saros/E but missing for Saros/I. But I couldn't find the logic that displays it in Saros/E (as no cancellation listeners were registered).

**Comment by srossbach** [comment]
This dialog is part of Eclipse logic. It is displayed when the Eclipse JOB is not successful.

https://github.com/saros-project/saros/blob/master/eclipse/src/saros/ui/eventhandler/NegotiationHandler.java

**Comment by tobous** [comment]
Pushed  the PR and added you as a reviewer. I flipped the calls to improve the readability.

**Comment by tobous** [comment]
> Pushed the PR and added you as a reviewer. I flipped the calls to improve the readability.

Damn, wrong window. xD Whatever.

> This dialog is part of Eclipse logic. It is displayed when the Eclipse JOB is not successful.
> 
> https://github.com/saros-project/saros/blob/master/eclipse/src/saros/ui/eventhandler/NegotiationHandler.java

Thanks for the hint. I will have a look at it.

**Comment by srossbach** [comment]
https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/runtime/UIMonitoredJob.java#L42

Checking the return status would be helpful.

**Comment by srossbach** [comment]
Changed the title to reflect the real issue.

**Comment by tobous** [comment]
> https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/runtime/UIMonitoredJob.java#L42
> 
> Checking the return status would be helpful.

Well, this is partly handled in the [NegotiationHandler](https://github.com/saros-project/saros/blob/master/intellij/src/saros/core/ui/eventhandler/NegotiationHandler.java#L135). But at that point, the error messages were already swallowed up by other parts of the negotiation (I think). And just stating that an error occurred is not really helpful. I would like to show the error message, especially in this case, where the error is fixable by the user.

But I still haven't gotten though all of the implementation. There are a lot of layers and it's kind of confusing.

**Comment by srossbach** [comment]
Displaying a popup which is visible for a few seconds, I would not call this "handled". And no it is not swallowed. Look at the code for case ERROR. It creates and Status of type error and that's it. Just evaluate the Status at the line I linked and display a Dialog afterwards if the status is <> OK.

**Comment by tobous** [comment]
> Displaying a popup which is visible for a few seconds, I would not call this "handled".

I disagree. In my opinion, notifications are the best way to inform the user. This is also specified in the [IntelliJ plugin developlment guidelines](https://www.jetbrains.org/intellij/sdk/docs/user_interface_components/notifications.html#notifications). And, from my experience working with IntelliJ, the vast majority of user updates (including error messages) are indeed displayed as notifications.

In IntelliJ, the notification popups are more than just short popups. They are colored according to the level (info, warn, and error). Futhermore, the user always has an indicator displaying the highest notification level that occurred (until it's dismissed) and a history over all notifications, giving them the change to continue interacting with the IDE without loosing access to the error message.

**Comment by srossbach** [comment]
Ah ok. I thought the notifications are equal to our custom made stuff we use in Eclipse (no history, only displayed for a few seconds).

**Comment by tobous** [comment]
Ok, the current Saros/I implementation of the status handling needs work. It was copied from the Eclipse counterpart but was left in an unfinished state that suggests more functionality than actually was implemented.

From looking at the Saros/E implementation in [NegotiationHandler](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/ui/eventhandler/NegotiationHandler.java#L80), it seems like the states were handled differently (some in the class and some externally) as the used Eclipse functionality already handles some cases ("ok", "cancel", and "error").

But, in IntelliJ, this separation does not really make sense. Still, it was *kind of* copied by adding a new [Status class](https://github.com/saros-project/saros/blob/master/intellij/src/saros/core/monitoring/Status.java) that is just a stub without an internal state.

The question now is whether the status should be handled internally (in the `run(Monitor)`) or externally by the caller. I am not sure which one is preferable.

@srossbach Any opinions on the matter?

**Comment by srossbach** [comment]
I guess you can just throw away all this UIMonitored custom made stuff and use IntelliJ API instead.

See 

https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/ui/wizards/AddProjectToSessionWizard.java#L608

I wrote the IProgressMonitor adapter for this, however it did not touch the already self made functionality.

**Comment by tobous** [comment]
I created a temporary fix for this with #680. I want to release Saros/I 0.2.0 and the real fix replacing the implementation would be too much work for right now. But I will fix it for the release after that.
